### PR TITLE
[WIP] Make FSDPv1 only perform cat() during last microbatch backward() within FlattenParamsWrapper 

### PR DESCRIPTION
**Pull request by chrisxcai**
```
sudo rm -r /tmp/chriscai/xldumps/* & CUDA_LAUNCH_BLOCKING=1 PYTHONPATH=~/benchmark/fairscale_repos/fairscale/ CRYPTOGRAPHY_OPENSSL_NO_LEGACY=1 torchrun --master_port 1024 --nproc_per_node=8 train.py --dump_dir /tmp/chriscai/xldumps --model_parallel_size 2 --pipeline_parallel_size 1 --seq_len=1024 --gpu_check_level=-1 --steps=10 --log_all_steps=True --profile_freq=10 --dump_profile_traces=True --profile_with_stack=True --use_microbatching=True --model.n_layers=1 --reshard_after_forward=False --batch_size=8 --model.efficient_attn=flash --model.attn_bias_type=causal --model=small --mem_snapshot_stop_step 5
```

trace before change (SplitWithSizesBackward triggered every microbatch per FSDP module):
https://fburl.com/perfdoctor/qdt32ibh

trace with applied change (SplitWithSizesBackward triggered only in last microbatch per FSDP module):
https://www.internalfb.com/intern/perfdoctor/trace_view?filepath=tree/traces/trace.229652302632210.json.gz&bucket=acadia

### CODE REVIEW

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py [hunk 1]
```diff
@@ -338,6 +338,11 @@ class FullyShardedDataParallel(nn.Module):
             rank 0 and return empty dict non-rank 0, which allow FullyShardedDataParallel to
             skip the GPU -> CPU copy on non-rank 0 altogether and prevent OOM.
             Default: False
+        optimize_backward_concat (bool):
+            If True, only let backward pass propagate to self.params, which will
+            invoke the _post_backward_hook() and concat() op, when self._require_backward_grad_sync
+            is True (e.g. last microbatch)
+            NOTE: this likely will incur more GPU memory usage
```
Comment (awgu) [inline fairscale/nn/data_parallel/fully_sharded_data_parallel.py:345]: Could you explain why there will be more GPU memory usage?

File: fairscale/nn/misc/flatten_params_wrapper.py [hunk 2]
```diff
@@ -364,18 +379,60 @@ def _unflatten_params(self, external_data: Optional[List[Optional[Tensor]]] = No
...
+    # The post backward hook used to accumulate fp32 gradients
+    def _grad_accumulation_hook(
+        self,
+        grad,
+        param_index,
+    ):
+        if self.fp32_grads[param_index] is None:
+            self.fp32_grads[param_index] = grad.to(torch.float32)
+        else:
+            self.fp32_grads[param_index].add_(grad.data)
```
Comment (awgu) [inline fairscale/nn/misc/flatten_params_wrapper.py:391]: nit: I think `grad.data` can just be `grad` (save one `aten.detach` call)

**Review by awgu — APPROVED** [review]
This approach makes sense to me!

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py (hunk 1 above)
Comment (chrisxcai) [reply fairscale/nn/data_parallel/fully_sharded_data_parallel.py:345]: hi @awgu, currently by testing results it shows the GPU memory overhead could be non-trivial (20% of 80G), we will follow up on reducing the memory usage
Screenshot 2024-05-15 at 10 39 19 AM
Screenshot 2024-05-15 at 10 40 18 AM
